keyboards/planck/rev7/matrix.c

```diff
@@ -111,13 +112,40 @@ bool matrix_scan_custom(matrix_row_t current_matrix[]) {
     return changed;
 }
 
-uint8_t encoder_quadrature_read_pin(uint8_t index, bool pad_b) {
-    pin_t pin = pad_b ? B13: B12;
-    gpio_set_pin_input_high(pin);
-    gpio_write_pin_low(matrix_row_pins[index]);
-    wait_us(10);
-    uint8_t ret = gpio_read_pin(pin) ? 1 : 0;
-    gpio_set_pin_input_low(matrix_row_pins[index]);
-    gpio_set_pin_input_low(pin);
-    return ret;
+#if defined(ENCODER_ENABLE) || defined(ENCODER_MAP_ENABLE)
+
+#if !defined(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY)
+#   define PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY 20
+#endif
+
+void encoder_driver_task(void) {
+
+    // set up C/rows for encoder read
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        gpio_set_pin_output(matrix_row_pins[i]);
+        gpio_write_pin_high(matrix_row_pins[i]);
+    }
+
+    // set up A & B for reading
+    gpio_set_pin_input_high(B12);
+    gpio_set_pin_input_high(B13);
+
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        gpio_write_pin_low(matrix_row_pins[i]);
+        wait_us(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY);
+        encoder_quadrature_handle_read(i, palReadPad(GPIOB, 12), palReadPad(GPIOB, 13));
+        gpio_write_pin_high(matrix_row_pins[i]);
+    }
+
+    // revert A & B to matrix state
+    gpio_set_pin_input_low(B12);
+    gpio_set_pin_input_low(B13);
+
+    // revert C/rows to matrix state
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        gpio_set_pin_input_low(matrix_row_pins[i]);
+    }
+
 }
+
+#endif // ENCODER_ENABLE || ENCODER_MAP_ENABLE
```

Comment:
Given the custom matrix implementation, I guess that something like this should work:
```suggestion
#if !defined(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY)
#   define PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY 20
#endif

void encoder_quadrature_init_pin(uint8_t index, bool pad_b) {
    // encoder_init() is called before matrix_init(), so the pin init needs to be done here.
    pin_t col_pin = pad_b ? B13 : B12;
    gpio_set_pin_input_low(col_pin);
    gpio_set_pin_input_low(matrix_row_pins[index]);
}

uint8_t encoder_quadrature_read_pin(uint8_t index, bool pad_b) {
    pin_t col_pin = pad_b ? B13 : B12;
    gpio_set_pin_high(col_pin);
    wait_us(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY);
    uint8_t ret = gpio_read_pin(matrix_row_pins[index]) ? 0 : 1;
    gpio_set_pin_input_low(col_pin);
    return ret;
}
```
(Although the matrix code uses `gpio_set_pin_low()` instead of `gpio_set_pin_input_low()` for inactive columns; maybe it should be used here too, then the encoder pin init code in `matrix_init_custom()` needs to be changed to match.)

It could be possible to avoid the multiple settle delays by reading the state of all encoders at once, but with the current API it would require either making some assumptions about the order of `encoder_quadrature_read_pin()` calls (and in the existing code the order of calls for pads A and B is actually unspecified), or overriding `encoder_driver_task()`.

And this implementation of `encoder_quadrature_init_pin()` might still not be enough, because it does not init the other column pins, so they might interfere with the `encoder_quadrature_read_pin()` calls done during `encoder_quadrature_post_init()` (that's unlikely though, because it would require that some keys are held down during startup).  Doing things in a 100% correct way may require moving the matrix pin init into `keyboard_pre_init_kb()`, then `encoder_quadrature_init_pin()` may be empty (the default implementation still needs to be overridden though).

Reply:
@sigprof Thank you for your review and sample code. 

I may do not understand your sample code, I tried use your code on my local env and but does not work correct. so I will try find other way as reference to your sample code, and I need a time more.

Reply:
Note: does not work code.

```suggestion
#if defined(ENCODER_ENABLE) || defined(ENCODER_MAP_ENABLE)

void encoder_quadrature_init_pin(uint8_t index, bool pad_b) {
    // encoder_init() is called before matrix_init(), so the pin init needs to be done here.
    pin_t col_pin = pad_b ? B13 : B12;
    gpio_set_pin_input_low(col_pin);
    gpio_set_pin_input_low(matrix_row_pins[index]);
}
uint8_t encoder_quadrature_read_pin(uint8_t index, bool pad_b) {
    pin_t col_pin = pad_b ? B13 : B12;
    gpio_set_pin_input_high(col_pin); // If using `gpio_set_pin_high`, compile error. so i did change to this.
    wait_us(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY);
    uint8_t ret = gpio_read_pin(matrix_row_pins[index]) ? 0 : 1;
    gpio_set_pin_input_low(col_pin);
    return ret;
}
#endif // ENCODER_ENABLE || ENCODER_MAP_ENABLE
```

Reply:
(And `gpio_set_pin_output(col_pin)` is needed before.)

Reply:
And your code which switches the row pins to the output mode somewhat defeats the point of the custom matrix implementation (to avoid using the C13…C15 pins as outputs).  My suggestion tries to keep those pins in the input mode, as the custom matrix code does (but apparently writing the code in the GitHub comment box is too hard).

Reply:
@sigprof I did update code with blank encoder_quadrature_init_pin function. (already called gpio_set_pin_input_low for B12 and B13 on matrix_init_custom(void))

> My suggestion tries to keep those pins in the input mode, as the custom matrix code does 

I'm not confident. but updated. Is this commit expected code? https://github.com/qmk/qmk_firmware/pull/23967/commits/e34700f2a1adc9d44a70d63e496abd4aa6aed557

Reply:
> encoder_init() is called before matrix_init(), so the pin init needs to be done here.

That commit has blank encoder_quadrature_init_pin function. but looks no problem for me on using. Should that use `encoder_quadrature_init_pin`? and needs delete init code in `matrix_init_custom`?